internsctl.1
.TH INTERNSCTL 1 "AUGUST 2023"
.SH NAME
internsctl \- custom Linux command for various operations

.SH SYNOPSIS
.B internsctl
[\fIOPTION\fR]
[\fBCOMMAND\fR]
[\fIARGUMENTS\fR]

.SH DESCRIPTION
.B internsctl
is a custom Linux command that provides various operations for system information and user management.

.SH OPTIONS
.TP
.BR --help, -h
Show the help message for the command.
.TP
.BR --version
Show the version of the \fBinternsctl\fR command.
.SH COMMANDS 
The available \fIcommands\fR for the \fBinternsctl\fR command are as follows:
.TP
.BR cpu getinfo
Get CPU information of the server.
.TP
.BR memory getinfo
Get memory information of the server.
.TP
.BR user create  \fR\fIusername\fR
Create a new user with the given \fIusername\fR. The user can log in to the Linux system and access their home directory.
.TP
.BR user list
List all regular users present on the server.
.TP
.BR user list --sudo-only
List all users with sudo permissions on the server.
.TP
.BR file getinfo \fR\fIfile-name\fR [\fIoptions\fR]
Get information aboutr a file with the given \fIfile-name\fR. Additional \fIoptions\fR can be provided to obtain specific information about the  file.

.SH FILE INFORMATION OPTIONS 
The following \fIoptions\fR can be used with the \fBfile getinfo\fR command to obtain specific information about a file:
.TP
.BR --size, -s
Print the size of the specified file only.
.TP
.BR --permissions, -p
Print the file permissions of the specified file.
.TP
.BR --owner, -o
Print the owner of the specified file.
.TP
.BR --last-modified, -m
Print the last modified time of the specified file.

.SH EXAMPLES 
.TP
\fBExample 1:\fR Get CPU informtion
.PP
.nf
.B internsctl cpu getinfo
.fi
.TP
\fBExample 2:\fR Get memory information
.PP
.nf
.B internsctl memory getinfo
.fi
.TP
\fBExample 3:\fR Create a new user 
.PP
.nf
.B internsctl user create john
.fi
.TP
\fBExample 4:\fR List all regular users:
.PP
.nf
.B internsctl user list
.fi
.TP
\fBExample 5:\fR List all users with sudo permissions:
.PP
.nf
.B internsctl user list --sudo-only
.fi
.TP
\fBExample 6:\fR Get information about a file named "hello.txt":
.PP
.nf
.B internsctl file getinfo hello.txt
.fi

.SH EXAMPLES FOR FILE COMMAND WITH OPTIONS
.TP
\fBExample 1:\fR Get the size of the file "hello.txt":
.PP
.nf
.B nternsctl file getinfo --size hello.txt
.fi
.TP
\fBExample 2:\fR Get the permissions of the file "hello.txt":
.PP
.nf
.B internsctl file getinfo --permissions hello.txt
.fi
.TP
\fBExample 3:\fR Get the owner of the file "hello.txt":
.PP
.nf
.B internsctl file getinfo --owner hello.txt
.fi
.TP
\fBExample 4:\fR Get the last modified time of the file "hello.txt":
.PP
.nf
.B internsctl file getinfo --last -modified hello.txt
.fi

.SH AUTHOR
Written by YourName.
.SH COPYRIGHT
Copyright (c) YourName. This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent permitted by law. 



internsctl.sh
#!/bin/bash

#internsctl.sh

#Variables
COMMAND_NAME="internsctl"
COMMAND_VERSION="v0.1.0"

#Function to display help information
function show_help {
    echo "Usage: $COMMAND_NAME [OPTION]"
    echo "Options:"
    echo " --help                  Display this help and exit"
    echo " --version               Display version information and exit"
    echo ""
    echo "Commands:"
    echo " cpu getinfo              Get CPU information"
    echo " memory getinfo           Get memory information"
    echo " user create <username>   Create a new user"
    echo " user list                List regular users"
    echo " user list --sudo-only    List users with sudo permissions"
    echo " file getinfo <file-name> Get information about a file"
}

# Function to display version information
function show_version {
    echo "$COMMAND_NAME $COMMAND_VERSION"
}

function get_cpu_info {
   lscpu
}

# Function to handle memory getinfo
function get_memory_info {
   free
}

#Function to handle user create
function create_user {
     if [[ $# -ne 1 ]]; then
        echo "Error: Missing username."
        echo "Usage: internsctl user create <username>"
        exit 1
     fi

    # Add code here to create a new user with the provided username
    # for example: useradd "$1"
   sudo useradd "$1"
   echo "User $1 created successfully."
}

#Function to handle user list
function list_users {
   # check if the "--sudo-only" option is provided
   if [ "$1" = "--sudo-only" ]; then
      getent group sudo wheel | cut -d: -f4 | tr ',' '\n'
   else
      # execute the appropriate command to list regular users
      getent passwd | cut -d: -f1
   fi
}

# Function to handle file getinfo
function get_file_info {
   if [[ "$#" -lt 1 ]]; then
       echo "Error: Missing file name."
       echo "Usage: internsctl file getinfo <file-name> [options]"
       exit 1

   fi

   local file_name="$1"
   shift

   #check if any options are provided
    while [[ "$#" -gt 0 ]]; do
         case "$1" in
              --size|-s)
                 echo $(wc -c < "$file_name")
                 ;;
              --permissions|-p)
                  grep -o 'Access: .*' "$file_name" | awk '{print $2}'
                  ;;
              --owner|-o)
                  grep -o 'Owner: .*' "$file_name" | awk '{print $2}'
                  ;;
              --last-modified|-m)
                  grep -o 'Modify: .*' "$file_name" | awk '{print $2, $3}'
                  ;;
               *)
                  echo "Unknown option: $1"
                  exit 1
                  ;;
            esac
            shift
         done
}

# Main function to handle command arguments
function main {
    case "$1" in
        --help|-h)
           show_help
           ;;
        --version)
            show_version
            ;;
         cpu)
            get_cpu_info
            ;;
         memory)
            get_memory_info
            ;;
         user)
            case "$2" in
              create)
                 create_user "$3"
                 ;;
            list)
               list_users "$3"
               ;;
             *)
                echo "Unknown command: internsctl user $2"
                exit 1
                ;;
          esac
          ;;
      file)
          if [[ "$2" == "getinfo" ]]; then
             shift 2
             get_file_info "$@"
          else
             echo "Unknown command: internsctl file $2"
             exit 1

          fi
          ;;
      *)
        echo "Unknown command: $1"
        show_help
        exit 1
        ;;
  esac
}

#call the main function with command line arguments
main "$@"

